### PR TITLE
noto-fonts-cjk: use sparse checkouts for smaller downloads

### LINKED ISSUE #156180: Noto Fonts CJK fails to build
### Describe the bug
When building my NixOS configuration, `w5r4h4cn7i6yk3xjhhvd7zv28xbbjydj-noto-fonts-cjk-sans-2.004.drv` failed to build.
I’m using `nixos-unstable` as my nixpkgs branch.

<details>

<summary> Error logs (click to expand) </summary>

```none
error: builder for '/nix/store/wx4spfncy8jfxirzqr1izj242p3dgywv-source.drv' failed with exit code 1;
       last 10 log lines:
       >                                  Dload  Upload   Total   Spent    Left  Speed
       > 100   162  100   162    0     0    154      0  0:00:01  0:00:01 --:--:--   154
       > 100  413M    0  413M    0     0   323k      0 --:--:--  0:21:50 --:--:--  317k
       > unpacking source archive /build/9f7f3c38eab63e1d1fddd8d50937fe4f1eacdb1d.tar.gz
       >
       > gzip: stdin: unexpected end of file
       > tar: Unexpected EOF in archive
       > tar: Unexpected EOF in archive
       > tar: Error is not recoverable: exiting now
       > do not know how to unpack source archive /build/9f7f3c38eab63e1d1fddd8d50937fe4f1eacdb1d.tar.gz
       For full logs, run 'nix log /nix/store/wx4spfncy8jfxirzqr1izj242p3dgywv-source.drv'.
error: 1 dependencies of derivation '/nix/store/w5r4h4cn7i6yk3xjhhvd7zv28xbbjydj-noto-fonts-cjk-sans-2.004.drv' failed to build
error: 1 dependencies of derivation '/nix/store/bdnwibjnj6xwwf1hd85nshsjn2l9d9mn-fc-00-nixos-cache.conf.drv' failed to build
error: 1 dependencies of derivation '/nix/store/9cw9clp87vgxk768gfmnwirk5plz5s10-xserver.conf.drv' failed to build
error: 1 dependencies of derivation '/nix/store/krbb4hzw6ca5c9icqx3nwd5nncz54lx5-fontconfig-conf.drv' failed to build
error: 1 dependencies of derivation '/nix/store/a0k1z3aybp3y8bzad3g2b84f9n24rkbg-unit-display-manager.service.drv' failed to build
error: 1 dependencies of derivation '/nix/store/dqm0ygz05dn7xscdcpv1049q3n5i75s3-fontconfig-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/wbilm4sk2gkxgq3njar7bapwjbfswddh-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/dfzfg8ypkbh3jmfxf3l17cms5al50ac0-nixos-system-buggeryyacht-nixos-22.05.20220119.6d82152.drv' failed to build
```

</details>

### Steps To Reproduce
Build the flake `github:schuelermine/nixos-configuration/81c34993dff48950d17ab476f4a09f8f39596001`

### Expected behavior
The system is buildt

### Notify maintainers

@mathnerd314
@emilazy

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```none
0 /etc/nixos $ nix run nixpkgs#nix-info -- -m
 - system: `"x86_64-linux"`
 - host os: `Linux 5.16.0, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.6.0pre20211217_6e6e998`
 - channels(root): `"nixos, nixpkgs"`
 - channels(anselmschueler): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by midchildan**
###### Motivation for this change

Fixes #156180. There were problems with building Noto CJK including on Hydra presumably because the size of the source archive was too large. This PR fixes this problem by fetching individual files instead of the whole archive.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Tested using a NixOS test in https://github.com/NickCao/nixpkgs/commit/2cb96de17b9dad600c7bb1a2a14e0a4a6f83933c to check that it renders correctly:

screen

### CODE REVIEW

File: pkgs/data/fonts/noto-fonts/default.nix [hunk 1]
```diff
@@ -96,6 +97,23 @@ let
         maintainers = with maintainers; [ mathnerd314 emily ];
       };
     };
```
Comment (Mathnerd314) [inline pkgs/data/fonts/noto-fonts/default.nix:99]: > Tested using a NixOS test in NickCao@2cb96de to check that it renders correctly

I think you might as well add the test to this PR, easier to keep track of that way.

According to [the manual](https://nixos.org/manual/nixpkgs/stable/#ssec-nixos-tests-linking) it should be
```
  passthru.tests = {
    noto-fonts = nixosTests.noto-fonts;
  };
```

**Review by Mathnerd314 — APPROVED** [review]
Otherwise LGTM.

File: pkgs/data/fonts/noto-fonts/default.nix (hunk 1 above)
Comment (midchildan) [reply pkgs/data/fonts/noto-fonts/default.nix:102]: I cherry-picked the commit into this PR.

@NickCao Are you okay with https://github.com/NixOS/nixpkgs/pull/156342/commits/7e0ff044a40423b33bafbc8046a6e29692200698 being merged?

**Comment by jtojnar** [comment]
This will make updating even harder. Perhaps there should be an update script.

**Comment by Mathnerd314** [comment]
> Perhaps there should be an update script.

I don't think it's worth the effort to automate running nix-prefetch-url a dozen times. Noto updates are rare, 6 months to a year and a half between updates. Plus since they are talking about splitting the repo into two, an update script is likely to break.

But maybe there is some update script toolkit that makes writing them simple? That would be nice.

**Review by LunNova — APPROVED** [review]

**Review by Mathnerd314 — APPROVED** [review]

File: nixos/tests/noto-fonts.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import ./make-test-python.nix ({ pkgs, lib, ... }: {
+  name = "noto-fonts";
+  meta = {
+    maintainers = with lib.maintainers; [ nickcao ];
```
Comment (SuperSandro2000) [inline nixos/tests/noto-fonts.nix:4]:
```suggestion
    maintainers = with lib.maintainers; [ midchildan ];
```

File: pkgs/data/fonts/noto-fonts/default.nix (hunk 1 above)
Comment (NickCao) [reply pkgs/data/fonts/noto-fonts/default.nix]: That's ok (sorry for the late response)

File: nixos/tests/noto-fonts.nix (hunk 2 above)
Comment (midchildan) [reply nixos/tests/noto-fonts.nix:4]: I should've mentioned this in a top level comment, but this is actually correct. The tests originate from @NickCao's code in https://github.com/NixOS/nixpkgs/pull/156080#issuecomment-1019016416. The relevant discussion is in https://github.com/NixOS/nixpkgs/pull/156342#discussion_r790330641.

Comment (NickCao) [reply nixos/tests/noto-fonts.nix:4]: I would be happy to co-maintain this, feel free to add yourself.

Comment (midchildan) [reply nixos/tests/noto-fonts.nix]: Thanks, added.

**Pull request merged**